driver: ethernet: Introduce ETH_DRIVER_RAW_MODE option

pdgendt opened the pull request:
Add a `Kconfig` symbol to allow building ethernet device drivers without an L2 layer.

~Currently this opens a can of worms with `NET_TEST` circular dependencies.~

~Depends on #69385~

pdgendt committed:
drivers: ethernet: Introduce ETH_DRIVER_RAW_MODE option

Add a Kconfig symbol to allow building ethernet device driver without
an L2 layer.

Signed-off-by: Pieter De Gendt <pieter.degendt@basalte.be>

jukkar commented on drivers/ethernet/Kconfig:
```diff
@@ -5,12 +5,20 @@
 	bool "Ethernet drivers"
-	default y
-	depends on NET_L2_ETHERNET
-	depends on !NET_TEST
```
The purpose of the `NET_TEST` is to prevent Ethernet drivers to be taken into use when building network tests, as those use simulated Ethernet drivers and are self contained. Using real Ethernet driver will just cause problems in the tests if there are extra network interfaces created. I don't think we should remove `NET_TEST` here.

pdgendt committed:
include: net: net_if: Re-use instance macro for NET_DEVICE_INIT

Re-use a single macro for both NET_DEVICE_INIT and
NET_DEVICE_INIT_INSTANCE.

Signed-off-by: Pieter De Gendt <pieter.degendt@basalte.be>

pdgendt committed:
include: net: ethernet: Add ETH_NET_DEVICE_INIT_INSTANCE

Add a macro variant for ethernet devices with instances.

Signed-off-by: Pieter De Gendt <pieter.degendt@basalte.be>

pdgendt committed:
drivers: ethernet: adin2111: Use ethernet init macro

Use the ethernet specific init macro.

Signed-off-by: Pieter De Gendt <pieter.degendt@basalte.be>

pdgendt committed:
drivers: ethernet: Build ethernet drivers with ETH_DRIVER_RAW_MODE

Allow building ethernet drivers without NET_L2_ETHERNET config symbol.

Signed-off-by: Pieter De Gendt <pieter.degendt@basalte.be>

pdgendt committed:
tests: drivers: build_all: ethernet: Build without L2

Limit the number of files built when building ethernet device drivers.
Use the ETH_DRIVER_RAW_MODE to provide weak funtions for upper layer
handling.

Signed-off-by: Pieter De Gendt <pieter.degendt@basalte.be>

pdgendt committed:
drivers: ethernet: nxp_s32_gmac: Imply MDIO

The ethernet driver will most likely require the MDIO device to be
enabled.

Signed-off-by: Pieter De Gendt <pieter.degendt@basalte.be>

jukkar commented:
Only reviewed the last three commits, they LGTM. Perhaps makes sense to wait #69385 to be merged before removing draft status.
